### ISSUE TITLE
[feature]: allow settlement of an invoice to yourself without routing the payment

### ISSUE THREAD

**Issue by AndySchroder** [lightningnetwork/lnd#7419]
**Describe the solution you'd like**

Allow settlement of an invoice to yourself without routing the payment through the lightning network. There are many cases where you'd want to be able to accept a payment from yourself. Currently people seem mostly interested in doing this for re-balancing channels, but there are cases where you may want to do so without re balancing channels.

**Describe alternatives you've considered**

1. Running 2 nodes: total waste of resources and complexity. Requires periodic channel re-balancing because of https://github.com/lightningnetwork/lnd/issues/7417.
2. Routing payments through other nodes using `allow_self_payment`: Works, but causes unnecessary fees, capacity requirements, latency, network traffic, channel unbalancing, etc..
3. Creating a private channel back to yourself on the same node: Can't do this, and even if you could https://github.com/lightningnetwork/lnd/issues/7417 probably would make this not work anyway.

**Additional context**

Anyone trying to use their own service can't. For example, a user of Distributed Charge - EV (http://andyschroder.com/DistributedCharge/) can't use their own charging station.

**Comment by AndySchroder** [comment]
Another use case for this is the LND Accounts feature (https://docs.lightning.engineering/lightning-network-tools/lightning-terminal/accounts) . Currently invoices are not able to be paid between two accounts using the same LND node.

**Comment by AndySchroder** [comment]
Putting this in a greater context on why this should be a standard feature: On all computers we have the loopback address, 127.0.0.1, which can be used to reach ourselves with internet protocol. The same functionality should exist for lightning.

**Comment by AndySchroder** [comment]
You need to set `--allow-circular-route` on your node to allow htlc forwards that arrive and depart on the same channel. This is a setting for lnd.

You need to set `--allow_self_payment` when sending a payment to allow sending a circular payment to self using the RPC.

I believe that the difference is that `--allow-circular-route` is for any payment that your node routes, regardless of whether you are sending it or someone else is. `--allow_self_payment` is for a payment you are sending.

**Comment by AndySchroder** [comment]
The other thing is that `--allow_self_payment` is something you can do without a circular route if you have the payment going in and out different channels (3 hops minimum). `--allow-circular-route` allows you to make a self payment with just 2 hops.

**Comment by MichaelAntonFischer** [comment]
Hi,
I find it a bit wild that this feature is still not implemented. It is very common that users have several services like BTCPayServer and LNbits connected to the same node. It should really be possible to send payments from one such system to the other without routing through the lightning network.

I constantly find myself building workarounds for this missing feature.

**Comment by thespielplatz** [comment]
@MichaelAntonFischer 
As long as the feature is not implemented you could use lndhub as account layer (connected to the node) and connect lnbits & btcpay to lndhub

**Comment by MichaelAntonFischer** [comment]
This is precisely what I mean with building workarounds.

**Comment by thespielplatz** [comment]
Ok - it seems I am missing a point? - I thought with LNDHub as middleware-layer you had todo the workaround only one time. At least in my case, I setup up LNDHub one time and selfpayment problem gone.

**Comment by MichaelAntonFischer** [comment]
No. LNbits doesn't natively support LND Hub as funding source. LNBits has LND hub built in, so this would be a workaround, but really it is annoying that all this has to happen.

LND surpassed CLN by adding Kubernetes natively, but what is the use of Kubernetes, when you then can't natively deploy a ton of services on top?

**Comment by AndySchroder** [comment]
"Lightning Terminal" provides the "groundwork" for a far superior middle ware than LNbits for providing accounts and other constraints because it uses an identical macaroon authentication scheme and gRPC with regular LND, as if the middle ware did not even exist. That way an application can interact with LND directly or through the "Lightning Terminal" middle ware with more limited abilities and the application won't even know it.

However, this low level bug makes it impractical to use.

Does anyone know of CLN has this same issue?

**Comment by MichaelAntonFischer** [comment]
I don’t think middleware shouldn’t be necessary at all. The only feature I really need is self_payments, but also limiting what grpc connected services can do and access are both must haves that I feel should be in lnd.

**Comment by weltitob** [comment]
Is this closed yet? Why has this still not been implemented ? @AndySchroder does it work in a newer version of lnd?

**Comment by kilrau** [comment]
Bump

**Comment by ziggie1984** [comment]
I am evaluating potential designs for this feature

**Comment by ziggie1984** [comment]
After evaluating the possible design choices, I would question whether this feature belongs in lnd’s core payment and invoice primitives.

If no HTLC is created, no channel state changes, and no payment is sent through the Lightning Network, then the operation is not really a Lightning payment. It is an internal accounting transfer between services that happen to use the same lnd node. In that model, an account layer above lnd still needs to track which service owns which balance, who is authorized to spend, and how internal debits and credits are reconciled.
Introducing invoices that are marked as paid even though no external Lightning payment occurred risks leaking application-level accounting semantics into lower-level lnd primitives. It could also make invoice history harder to reason about, since “settled” would no longer necessarily mean “paid through Lightning,” but could also mean “locally marked as settled by an
internal ledger.”

For that reason, I think this may be better handled by applications or account systems built on top of lnd. Those systems can expose invoice-like workflows to their users while maintaining the correct internal balance accounting, without changing the meaning of lnd’s core invoice and payment flows.

**Comment by ekzyis** [comment]
Thank you for looking into this @ziggie1984. Makes sense to me.

As a user, I'm sad, but as a developer, I applaud your decision and your communication.

**Comment by tlindi** [comment]
So issue seems to be structrular limitation of Lightning? And impossible (does such thing exist?) to solve correctly.

**Comment by ZZiigguurraatt** [comment]
Based on an offline discussion we think we are going to close this issue in favor of https://github.com/lightninglabs/lightning-terminal/issues/494 .

Generally the team felt that it was not the correct architecture to do anything inside of LND to handle self payments that don't reflect a real HTLC taken in and out of flight. Also, there is some concern of how an accountant would handle self payments that were made exclusively within LND.

Also, it's likely as lightning scales and is used by a large number of human and machine uses cases inside an organization on the same LND node, it is not going to make sense to give a macaroon directly from LND to the human or the machine. Instead, a more narrowly scoped balance (and possibly also narrowly scoped spending rates: https://github.com/lightninglabs/lightning-terminal/issues/583), controlled by a macaroon provided by the LND Accounts feature inside of litd (or some other middle ware) is likely the best and safest approach that end users should use LND through. Humans and machines should have their capabilities constrained more closely to what their use case of the underlying LND node actually is and LND Accounts in litd is an example of a tool that does such a thing.
